File: Sangho/Week08/도시분할계획.java
import java.util.*;
import java.io.*;

public class Main {

    static StringTokenizer st;

    static int N, M;

    static int[] parent;

    public static class Edge implements Comparable<Edge> {
        int start, destination, weight;

        Edge(int start, int destination, int weight) {
            this.start = start;
            this.destination = destination;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge o) {
            return this.weight - o.weight;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        Edge[] edges = new Edge[M];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int destination = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());

            edges[i] = new Edge(start, destination, weight);
        }

        int answer = kruskal(N, edges);

        System.out.println(answer);
    }

    public static int kruskal(int n, Edge[] edges) {
        Arrays.sort(edges);

        parent = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }

        int fatigue = 0;
        int bigcost = 0;
        int count = 0;

        for (Edge edge : edges) {
            if (find(edge.start) != find(edge.destination)) {

                fatigue += edge.weight;
                union(edge.start, edge.destination);
                bigcost = edge.weight;
                count++;

            }

            if (count == n - 1) {
                break;
            }
        }

        return fatigue - bigcost;
    }

    public static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    public static void union(int x, int y) {
        x = find(x);
        y = find(y);

        if (x != y) {
            parent[x] = y;
        }
    }
}

File: UnrequiredOne/Week08/도시분할계획.py
def union(a, b):
    a = find(a)
    b = find(b)
    if a == b:
        return

    if a < b:
        parent[b] = a
    else:
        parent[a] = b


def find(a):
    if parent[a] == a:
        return a
    parent[a] = find(parent[a])
    return parent[a]


N, M = map(int, input().split())

parent = [i for i in range(N+1)]
edges = []
answer = 0

for _ in range(M):
    A, B, C = map(int, input().split())
    edges.append((C, A, B))

edges.sort(key=lambda x: x[0])
max_cost = 0
for i in range(M):
    c, a, b = edges[i]
    if find(a) == find(b):
        continue
    else:
        union(a, b)
        answer += c
        max_cost = max(max_cost, c)
print(answer - max_cost)
